### ISSUE TITLE
[BUG] fvm runs commands twice on [Windows]

### ISSUE THREAD

**Issue by rydmike** [leoafarias/fvm#227]
## fvm command on Windows runs twice

On Windows when I run **fvm** it has for quite a while been running the given command twice:

Like so:
```
> fvm use dev --verbose
Project now uses Flutter: dev
Project now uses Flutter: dev

> fvm flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel dev, 1.27.0-8.0.pre, on Microsoft Windows [Version 10.0.19041.804], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.8.3)
[√] Android Studio (version 4.1.0)
[√] IntelliJ IDEA Community Edition (version 2020.3)
[√] VS Code (version 1.53.2)
[√] Connected device (5 available)

• No issues found!
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel dev, 1.27.0-8.0.pre, on Microsoft Windows [Version 10.0.19041.804], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.8.3)
[√] Android Studio (version 4.1.0)
[√] IntelliJ IDEA Community Edition (version 2020.3)
[√] VS Code (version 1.53.2)
[√] Connected device (5 available)

• No issues found!

>  fvm version
1.3.7
1.3.7
````

**Environment:**
 - OS: Windows10
 - Shell: PowerShell 7.1.2
 - FVM: 1.3.7

**Additional context**
Sorry, not sure with which version of fvm it started happening, but I think I have seen it at least for 2 months. Just never got around to reporting it. 

Everything works, but the commands always running twice is a bit annoying to say the least.

**Comment by Tokenizer94** [comment]
I have the same issue.

**Comment by rydmike** [comment]
@Tokenizer94 Thanks for sharing that, feels good to know I'm not seeing double! 😃

**Comment by leoafarias** [comment]
I am not sure what is happening here. I was not able to replicate with the latest stable version of the dart sdk. I have been having lots of different issues with the beta version.

Could you guys confirm which version of the Dart sdk is being used?

**Comment by rydmike** [comment]
Happens for me with Dart SDK **2.10.5** on channel stable, ask version, get answer twice:

```
>fvm flutter --version
Flutter 1.22.6 • channel stable • https://github.com/flutter/flutter.git 
Framework • revision 9b2d32b605 (5 weeks ago) • 2021-01-22 14:36:39 -0800
Engine • revision 2f0af37152
Tools • Dart 2.10.5
Flutter 1.22.6 • channel stable • https://github.com/flutter/flutter.git 
Framework • revision 9b2d32b605 (5 weeks ago) • 2021-01-22 14:36:39 -0800
Engine • revision 2f0af37152
Tools • Dart 2.10.5
```

and also channel dev with Dart SDK **2.13.0**
Switching:

```
>fvm use dev
Project now uses Flutter: dev
Project now uses Flutter: dev
```

Twice as usual and then check version:

```
>fvm flutter --version
Flutter 1.27.0-8.0.pre • channel dev • https://github.com/flutter/flutter.git
Framework • revision b7d4806243 (10 days ago) • 2021-02-19 09:22:45 -0800    
Engine • revision 6993cb229b
Tools • Dart 2.13.0 (build 2.13.0-30.0.dev)
Flutter 1.27.0-8.0.pre • channel dev • https://github.com/flutter/flutter.git
Framework • revision b7d4806243 (10 days ago) • 2021-02-19 09:22:45 -0800    
Engine • revision 6993cb229b
Tools • Dart 2.13.0 (build 2.13.0-30.0.dev)
```


It happens with both VS-Code and IntelliJ, using Microsoft **PowerShell 7.1.2** in both.

**Comment by leoafarias** [comment]
The `fvm version` command does a standard print. Very weird that its showing up twice. I will try to dig in a bit more later, I have been trying on my windows VM as I am away from my windows machine for the next few days.

**Comment by ofury** [comment]
I`m having the same issue with Dart SDK version: 2.12.0:
```
❯ dart --version
Dart SDK version: 2.12.0 (stable) (Thu Feb 25 19:50:53 2021 +0100) on "windows_x64"


❯ fvm
Flutter Version Management: A cli to manage Flutter SDK versions.

Usage: fvm <command> [arguments]

Global options:
-h, --help       Print this usage information.
    --verbose    Print verbose output.

Available commands:
  config     Set configuration for FVM
  flutter    Proxies Flutter Commands
  install    Installs Flutter SDK Version
  list       Lists installed Flutter SDK Version
  releases   Lists Flutter SDK releases.
  remove     Removes Flutter SDK Version
  use        Which Flutter SDK Version you would like to use
  version    Prints the currently-installed version of FVM

Run "fvm help <command>" for more information about a command.
Flutter Version Management: A cli to manage Flutter SDK versions.

Usage: fvm <command> [arguments]

Global options:
-h, --help       Print this usage information.
    --verbose    Print verbose output.

Available commands:
  config     Set configuration for FVM
  flutter    Proxies Flutter Commands
  install    Installs Flutter SDK Version
  list       Lists installed Flutter SDK Version
  releases   Lists Flutter SDK releases.
  remove     Removes Flutter SDK Version
  use        Which Flutter SDK Version you would like to use
  version    Prints the currently-installed version of FVM

Run "fvm help <command>" for more information about a command.

```

**Comment by iqfareez** [comment]
Same goes with me, maybe after I upgrade Dart to `2.12.x`

**Comment by Tokenizer94** [comment]
Still no fix? It's so annoying.

**Comment by kuhnroyal** [comment]
> Still no fix? It's so annoying.

This an open source project, you are welcome to investigate and fix. So annoying.

**Comment by cezaryjk** [comment]
Hi, 
First of all, I would like to thank you for a useful tool.

I think I found a culprit. 
This inconvenience is caused by a redundant call for command in FVM.BAT
Attached please find my suggested corrections (mark as CK)

Thank you
Best regards
Cezary
[fvm.bat.txt](https://github.com/leoafarias/fvm/files/6097019/fvm.bat.txt)

**Comment by cezaryjk** [comment]
Hi Leo, 
You're absolutely right: it's Pub's fault. 
It seems that only specific Windows versions suffer from this. Mine is Windows 10.

I've tried with previous FVM version. Unsuccessfully. Pub keeps on recreating the same batch file.   

It turned out they changed "quite good" script to current one: 
[https://www.gitmemory.com/issue/dart-lang/pub/2670/703804339](url) 
I just posted a comment in closed issue in Dart.Pub  [https://github.com/dart-lang/pub/issues/2875#issuecomment-796731214](url) 

Fortunately, my script will work until snapshot upgrade  :-) 

BR 
Cezary
PS. I had to add "flutter" before "pub" because of missing PATH to Dart-SDK\bin folder.  

[fvmck.bat.txt](https://github.com/leoafarias/fvm/files/6123292/fvmck.bat.txt)

**Comment by leoafarias** [comment]
@cezaryjk thanks for the info

@rydmike @ofury @Tokenizer94 Could you guys try one of the test cases to see if its also happening?

Case 1: (flutter_asset_generator)
- dart pub global activate flutter_asset_generator.
- Find an example that has assets defined in pubspec. (e.g. https://github.com/CaiJingLong/flutter_resource_generator/tree/master/example)
- fgen --no-preview --no-watch.

Case 2: (ff_annotation_route)
- dart pub global activate ff_annotation_route.
- Use the example that has routes annotated. (e.g. https://github.com/fluttercandies/ff_annotation_route/tree/master/example)
- ff_route.

**Comment by Tokenizer94** [comment]
I found the problem.
I think this happens if you use this command once: "flutter pub global activate fvm"
even after you deactivate FVM using "flutter pub global deactivate fvm" , this issue won't be fixed!

I did this and the problem solved:
1- Remove flutter-SDK completely.
2- Remove the FVM folder and all of the installed versions in it.
3- Remove all paths related to flutter-SDK and FVM in the PATH (environment variables).
4- Install dart-SDK (Not flutter-SDK). Then restart your PC.
5- Make sure the dart-SDK path is added to the PATH. Then run this command "pub global activate fvm".
6- Install any version using fvm and you will see commands only once, not twice.

**Comment by cezaryjk** [comment]
Hi guys,

I’m terrible sorry but it looks like it was false positive. 

@Tokenizer94, you were right, the problem was caused by my „complex” environment: I have both Dart and Flutter SDK installed. And yet 1.22.6 and 2.0.1. 
It disappeared just after Dart SDK was “promoted” in PATH. 
 
![obraz](https://user-images.githubusercontent.com/12769666/110958891-7266eb00-834d-11eb-9ed9-a8715ae17324.png)
 

But still my version of FVM.BAT works somehow better, giving single course in any paths order in PATH. Probable because launching DART.BAT instead of DART.EXE breaks IF… GOTO commands.   

Thank you very much
Best regards
Cezary

**Comment by rydmike** [comment]
Thanks guys for all the above suggestions, I will try to find time to investigate and test them during the weekend too.

I actually don't have the Dart SDK installed at all separately, I did a year ago or so, but removed it due to some conflicts I had with it at one point. I do keep a separate Flutter SDK stable installed, plus latest version of all channels with fvm, including stable, and occasionally I might have a few selected specific version in use with fvm as well.

I've always had Flutter SDK and fvm installed with all channels, but did in the past not have any issue with fvm running commands twice.

**Comment by ofury** [comment]
I also have a separate Flutter SDK stable installed along with FVM.

I was able to fix the issue following the steps suggested above.

1. removed dart-sdk, FVM and environment variables related to Flutter SDK, Dart and FVM 
2. installed dart-sdk, activated FVM 
3. added to PATH separately installed Flutter SDK 

Thanks guys

**Comment by rydmike** [comment]
I was still having this issue and did not really know what I should try, the documentation on what paths to where that are needed for what, and in what order, is **abysmal in Flutter**. Well I finally decided to experiment as well and see what I can do about it.

I **don't** have the separate **Dart SDK installed**, only the one that comes bundled with Flutter. I would prefer to not install the separate Dart SDK again, as it should not be required if you don't make pure dart apps, and even they can apparently be made with just the version that comes bundled with the Flutter SDK nowadays, or so I seem recall reading somewhere. In the past this was not so, I had Dart SDK installed a long time ago, but it was de-installed even before I installed fvm.

My env path, the parts related to Flutter, were just these in this order:

```
C:\src\flutter\bin
C:\src\flutter\.pub-cache\bin
C:\src\flutter\bin\cache\dart-sdk\bin
```

I have no idea if I need them all, but if I look into what `C:\src\flutter\bin` contains, like the bat for **flutter**, I can deduce that it is needed, and quite correctly removing it kills the *flutter* command.

The same folder contains a **dart.bat** that seem to be able to direct the `dart` command to `C:\src\flutter\bin\cache\dart-sdk\bin` folder where it is found, so for that we might not need the `C:\src\flutter\bin\cache\dart-sdk\bin` path, but for all the other **Dart SDK** commands (pub, dartdevc, dart2js, etc...) in the folder we do need that path, so it has to be kept.

Then comes the question, what is path `C:\src\flutter\.pub-cache\bin` for? Do I need it? For what?

If I remove it and only use these two paths in this order:

```
C:\src\flutter\bin
C:\src\flutter\bin\cache\dart-sdk\bin
```

**Then the double execution of fvm stops!**
I still seem to be able to run **flutter** and **dart** sdk related command without **fvm**, and **fvm** seems to work too, at least in my initial tests.

So what do I miss by removing `C:\src\flutter\.pub-cache\bin`? Why was it there? No idea.  

I'm keeping it removed for now, but if I later notice any issues I might of course have to add it back and then get double runs of **fvm** again.

If anybody has a link to a doc that explains all the currently required env and path variables in a Flutter Windows installation, what they do and in what order they should/must be, that would be useful and interesting to learn more about.

**Comment by leoafarias** [comment]
@rydmike interesting. The pub-cache path is for the global pub commands so I am surprised fvm is running.

https://dart.dev/tools/pub/cmd/pub-global#running-a-script-from-your-path

However the difference from your setup and the docs is that the path for pub-cache is setup under the user directory.

I suggest adding it back as described in the docs and see if you still have the issue.

**Comment by rydmike** [comment]
OK I just noticed, I do still also have this in the path:

`C:\Users\<user>\AppData\Roaming\Pub\Cache\bin`
This folder contains:
```
detective.bat
devtools.bat
fvm.bat
stagehand.bat
```

Saw it before too, I just did not recognize what it was and did not look at what it contained then.  So there are thus now these three Flutter/Dart folders in the path, in this order:

```
C:\src\flutter\bin
C:\src\flutter\bin\cache\dart-sdk\bin
C:\Users\<user>\AppData\Roaming\Pub\Cache\bin;
```

**Comment by leoafarias** [comment]
That should be correct. Is flutter, dart and global packages.

**Comment by rydmike** [comment]
Indeed and seems to work well too. I just wonder where the path:

`C:\src\flutter\.pub-cache\bin`

Came from. I can't recall adding it myself, and removing it solved my double runs and so far I have not noticed any issues from having removed it.

**Comment by rydmike** [comment]
And the issue is back, no idea why. Getting those lovely double runs of fvm again. Pain royale this issue is... sigh. 😅

**Comment by leoafarias** [comment]
@rydmike can you tell me a bit about your environment, so I can try to replicate it, so I can try recreate it.

Did you install fim with flutter pub or dart? Which version are you currently using. I wonder if it has something to do with the logging itself, but haven’t been able to replicate it

**Comment by leoafarias** [comment]
@rydmike also I have made a lot of improvements and changes for 2.0 which is currently in dev
Do you want to give it a try to see if the issue happens on that version also?

`pub global activate fvm 2.0.0-dev.3`

**Comment by rydmike** [comment]
I originally installed it with flutter pub, as I did not have the separate Dart sdk installed, I still don't. 
I was running 1.3.8 above. Now trying 2.0.0-dev.3.

```
$fvm --version
2.0.0-dev.3
2.0.0-dev.3
```

Still runs twice as before.

**Comment by rydmike** [comment]
I tinkered with the order of the paths:

From:
```
C:\src\flutter\bin
C:\src\flutter\bin\cache\dart-sdk\bin
C:\Users\<user>\AppData\Roaming\Pub\Cache\bin
```

To:
```
C:\Users\<user>\AppData\Roaming\Pub\Cache\bin
C:\src\flutter\bin\cache\dart-sdk\bin
C:\src\flutter\bin
```

Now it runs just once again, no idea if this none standard order is bad for other Flutter/Dart commands though, at least flutter, dart and pub commands seemed to run, did not check others yet. OR if the issue will come back again, we will see in a day or too.

**BTW:** Is it safe to stay on version 2.0.0-dev.3?

**Comment by leoafarias** [comment]
I was able to replicate the issue. However, it seems that it is not an FVM problem. It is also happening with other global CLI.

Just ran a test with `very_good_cli` and the same thing is happening and is getting printed twice

**Comment by leoafarias** [comment]
> I tinkered with the order of the paths:
> 
> From:
> 
> ```
> C:\src\flutter\bin
> C:\src\flutter\bin\cache\dart-sdk\bin
> C:\Users\<user>\AppData\Roaming\Pub\Cache\bin
> ```
> 
> To:
> 
> ```
> C:\Users\<user>\AppData\Roaming\Pub\Cache\bin
> C:\src\flutter\bin\cache\dart-sdk\bin
> C:\src\flutter\bin
> ```
> 
> Now it runs just once again, no idea if this none standard order is bad for other Flutter/Dart commands though, at least flutter, dart and pub commands seemed to run, did not check others yet. OR if the issue will come back again, we will see in a day or too.
> 
> **BTW:** Is it safe to stay on version 2.0.0-dev.3?

Still WIP but I think it's fairly stable. To be honest  having someone beta test on windows would definitely not hurt.

**Comment by rydmike** [comment]
>Just ran a test with very_good_cli and the same thing is happening and is getting printed twice
Ok that is good to know, but what is actually causing it? Something we should report elsewhere?

> Still WIP but I think it's fairly stable. To be honest having someone beta test on windows would definitely not hurt.
I'll keep using it then! 😃

**Comment by leoafarias** [comment]
I am reviewing their docs, to try to pinpoint the problem.

**Comment by leoafarias** [comment]
Ok here is the issue. https://github.com/dart-lang/pub/issues/2934

**Comment by leoafarias** [comment]
For anyone else stumbling into this issue. Make sure `flutter/bin/cache/dart-sdk/bin/dart` is not in your PATH.
After you remove run `dart pub global activate fvm`

**Comment by rydmike** [comment]
Cool and thanks, at least we know why now and that fvm is not the issue! 💙
Yeah I was just going to say, might as well close this issue

**Comment by rydmike** [comment]
About removing the `flutter/bin/cache/dart-sdk/bin` as a path, just be aware that if you do remove it, you will not be able to run the commands in that folder directly. At least I could not, and I need those commands.

So far the order I mention above is still working OK for me.

Let's hope for a better final solution via https://github.com/dart-lang/pub/issues/2934.

Thanks @leoafarias for all you help and assist with this issue that was not even caused by **fvm**, much appreciated! 💙 👍🏻

**Comment by Levi-Lesches** [comment]
Hi guys! Author of https://github.com/dart-lang/pub/issues/2934 here (wow I'm famous 😄). Not sure why I didn't just do it then, but I submitted a PR now to remove the broken line from the Batch file, [as you guys noticed](https://github.com/leoafarias/fvm/issues/227#issuecomment-792263055). Hopefully it gets merged. I'm interested in how this passed all their tests since the latest release of Dart touted all the new `dart` commands, and the merge of `pub` into `dart pub` as its greatest feature. 

Anyway, do you guys happen to know why moving the PATH around helps? Using the `where` command, I found that there's only one file with the package I'm using, in `AppData\Roaming\Pub\Cache\bin`, so it shouldn't make a difference what the order of my PATH is. I think it's the specific version of Dart (from `\cache\dart-sdk` or `\bin`) that might be messing with things. I'll look into it more.

**Comment by Levi-Lesches** [comment]
Hey I figured something out: Not only does removing the second `dart pub` call fix the problem, but if you change the `dart %snapshot%` to `call dart %snapshot%` (and keep the second `dart pub` call), it works! I'm going to look into the documentation for the `CALL` command to figure this out.

**Comment by kzrfaisal** [comment]
QUICK HACK:
Open your fvm batch file, it could be at 
```
C:\Users\<user>\AppData\Local\Pub\Cache\bin\fvm.bat
or
C:\Users\<user>\AppData\Roaming\Pub\Cache\bin\fvm.bat
```

fvm.bat
```
@echo off
rem This file was created by pub v2.13.4.
rem Package: fvm
rem Version: 2.2.3
rem Executable: fvm
rem Script: main
if exist "C:\Users\kzrdf\AppData\Local\Pub\Cache\global_packages\fvm\bin\main.dart-2.13.4.snapshot" (
  dart "C:\Users\kzrdf\AppData\Local\Pub\Cache\global_packages\fvm\bin\main.dart-2.13.4.snapshot" %*
  rem The VM exits with code 253 if the snapshot version is out-of-date.
  rem If it is, we need to delete it and run "pub global" manually.
  if not errorlevel 253 (
    goto error
  )
  pub global run fvm:main %*
) else (
  pub global run fvm:main %*
)
goto eof
:error
exit /b %errorlevel%
:eof

```

Comment or remove `pub global run fvm:main %*` before the else block.
To comment just add `rem` before the line and finally your file will look like this
```
@echo off
rem This file was created by pub v2.13.4.
rem Package: fvm
rem Version: 2.2.3
rem Executable: fvm
rem Script: main
if exist "C:\Users\kzrdf\AppData\Local\Pub\Cache\global_packages\fvm\bin\main.dart-2.13.4.snapshot" (
  dart "C:\Users\kzrdf\AppData\Local\Pub\Cache\global_packages\fvm\bin\main.dart-2.13.4.snapshot" %*
  rem The VM exits with code 253 if the snapshot version is out-of-date.
  rem If it is, we need to delete it and run "pub global" manually.
  if not errorlevel 253 (
    goto error
  )
  rem pub global run fvm:main %*
) else (
  pub global run fvm:main %*
)
goto eof
:error
exit /b %errorlevel%
:eof

```
SAVE it and hopefully the issue will be resolved.

**Comment by Levi-Lesches** [comment]
@kzrfaisal, your solution will break the batch file in certain cases. Instead, change this line from 
```
dart "C:\Users\USERNAME\AppData\Local\Pub\Cache\global_packages\fvm\bin\main.dart-2.13.4.snapshot" %*
```
to
```
call dart "C:\Users\USERNAME\AppData\Local\Pub\Cache\global_packages\fvm\bin\main.dart-2.13.4.snapshot" %*
```

Now that https://github.com/dart-lang/pub/pull/3055 is approved, this issue will be solved when the commit reaches `stable`, and `dart pub global install` will include this fix. It's already been merged into `dart-lang/sdk`.
